File: 2-Graph Search, Shortest Path, and Data Structures/3-Graph Search/1-Breadth-First Search (BFS)/BFSApp.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import undirected_graph.AdjacencyList;

public class BFSApp {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        Scanner scanner = null;
        try {
            scanner = new Scanner(new File("undirected_graph_info.txt"));
            
            // Construct the graph
            AdjacencyList graph = new AdjacencyList();
            // Add the vertices
            int nVtx = Integer.parseInt(scanner.nextLine());
            for (int vtxID = 1; vtxID <= nVtx; ++vtxID) {
                graph.addVtx(vtxID);
            }
            // Add the edges
            while (scanner.hasNextLine()) {
                String[] ends = scanner.nextLine().split(" ");
                graph.addEdge(Integer.parseInt(ends[0]), Integer.parseInt(ends[1]));
            }

            // Find all the findable vertices starting from vertex #1 using BFS
            System.out.println("Findable vertices from vertex #1: " + graph.bfs(1)); // [1, 2, 3, 4, 5, 6]
            graph.clearExplored();

            // Find the length of the shortest path from vertex #1 to vertex #6 using BFS
            System.out.println("Length of the shortest path from vertex #1 to vertex #6: " + graph.shortestPath(1, 6)); // 3
            graph.clearExplored();

            // Finds the number of connected components of the graph
            System.out.println("Number of connected components of the graph: " + graph.numOfConnectedComponents()); // 1
        } catch (FileNotFoundException ex) {
            System.out.println("Cannot find the file.");
        } finally {
            if (scanner != null) {
                scanner.close();
            }
        }
    }

}

File: 2-Graph Search, Shortest Path, and Data Structures/3-Graph Search/1-Breadth-First Search (BFS)/bfs_app.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

__author__ = 'Ziang Lu'

from undirected_graph import AdjacencyList


def main():
    with open('undirected_graph_info.txt', 'rt') as f:
        # Construct the graph
        graph = AdjacencyList()
        # Add the vertices
        n_vtx = int(f.readline())
        for vtx_id in range(1, n_vtx + 1):
            graph.add_vtx(new_vtx_id=vtx_id)
        # Add the edges
        for line in f.readlines():
            ends = line.split(' ')
            graph.add_edge(end1_id=int(ends[0]), end2_id=int(ends[1]))

        # Find all the findable vertices starting from vertex #1 using BFS
        print('Findable vertices from vertex #1: %s' % graph.bfs(src_vtx_id=1))  # [1, 2, 3, 4, 5, 6]
        graph.clear_explored()

        # Find the length of the shorted path from vertex #1 to vertex #6 using
        # BFS
        print('Length of the shortest path from vertex #1 to vertex #6: %d' %
              graph.shortest_path(src_vtx_id=1, dest_vtx_id=6))  # 3
        graph.clear_explored()

        print('Number of connected components: %d' %
              graph.num_of_connected_components())


if __name__ == '__main__':
    main()
